File: software/examples/rust/example_callback.rs
use std::{error::Error, io, thread};
use tinkerforge::{ip_connection::IpConnection, ozone_bricklet::*};

const HOST: &str = "localhost";
const PORT: u16 = 4223;
const UID: &str = "XYZ"; // Change XYZ to the UID of your Ozone Bricklet.

fn main() -> Result<(), Box<dyn Error>> {
    let ipcon = IpConnection::new(); // Create IP connection.
    let o = OzoneBricklet::new(UID, &ipcon); // Create device object.

    ipcon.connect((HOST, PORT)).recv()??; // Connect to brickd.
                                          // Don't use device before ipcon is connected.

    // Create receiver for ozone concentration events.
    let ozone_concentration_receiver = o.get_ozone_concentration_receiver();

    // Spawn thread to handle received events. This thread ends when the `o` object
    // is dropped, so there is no need for manual cleanup.
    thread::spawn(move || {
        for ozone_concentration in ozone_concentration_receiver {
            println!("Ozone Concentration: {} ppb", ozone_concentration);
        }
    });

    // Set period for ozone concentration receiver to 1s (1000ms).
    // Note: The ozone concentration callback is only called every second
    //       if the ozone concentration has changed since the last call!
    o.set_ozone_concentration_callback_period(1000);

    println!("Press enter to exit.");
    let mut _input = String::new();
    io::stdin().read_line(&mut _input)?;
    ipcon.disconnect();
    Ok(())
}

File: software/examples/go/example_callback.go
package main

import (
	"fmt"
	"github.com/Tinkerforge/go-api-bindings/ipconnection"
	"github.com/Tinkerforge/go-api-bindings/ozone_bricklet"
)

const ADDR string = "localhost:4223"
const UID string = "XYZ" // Change XYZ to the UID of your Ozone Bricklet.

func main() {
	ipcon := ipconnection.New()
	defer ipcon.Close()
	o, _ := ozone_bricklet.New(UID, &ipcon) // Create device object.

	ipcon.Connect(ADDR) // Connect to brickd.
	defer ipcon.Disconnect()
	// Don't use device before ipcon is connected.

	o.RegisterOzoneConcentrationCallback(func(ozoneConcentration uint16) {
		fmt.Printf("Ozone Concentration: %d ppb\n", ozoneConcentration)
	})

	// Set period for ozone concentration receiver to 1s (1000ms).
	// Note: The ozone concentration callback is only called every second
	//       if the ozone concentration has changed since the last call!
	o.SetOzoneConcentrationCallbackPeriod(1000)

	fmt.Print("Press enter to exit.")
	fmt.Scanln()

}
